modules/st2-auto-form/modules/st2-form-text-field/directive.js

```diff
@@ -0,0 +1,196 @@
+              })
+              .data('st2FormTextField', true)
+              .on('focus blur', function() {
+                $container.toggleClass('st2-auto-form__field-container--focused');
+
+                if (isTouched) {
+                  $container.addClass('st2-auto-form__field-container--touched');
+                }
+
+                isTouched = true;
+              })
+
+              // Prevent textarea scroll bubbling
+              .on('mousewheel DOMMouseScroll', function(e) {
+                var delta = -e.wheelDeltaY || e.detail;
+
+                if (textarea.scrollHeight > textarea.clientHeight &&
+                     (delta < 0 && !textarea.scrollTop ||
+                      delta > 0 && textarea.scrollTop + textarea.clientHeight === textarea.scrollHeight)
+                   ) {
+                  e.preventDefault();
+                }
+              });
+
+            scope.$watch(function() {
+              return ctrl.$modelValue;
+            }, adjustAndValidate);
+
+            textarea.addEventListener('input', adjustAndValidate, false);
+
+            // Oninput in IE9 doesn't capture character deletion,
+            // so we use onkeyup to overcome this
+            if ('onpropertychange' in textarea && 'oninput' in textarea) {
+              textarea.addEventListener('keyup', boundAdjust, false);
+            }
+
+            scope.$on('$destroy', function() {
+              $clone.remove();
+            });
+
+            scope.$watch('minRows', function(minRows) {
+              $textarea.minRows = window.parseInt(minRows, 10);
+              boundAdjust();
+            });
+
+            scope.$watch('maxRows', function(maxRows) {
+              $textarea.maxRows = window.parseInt(maxRows, 10);
+              boundAdjust();
+            });
+
+            setTimeout(boundAdjust, 0);
+          }
+        };
+      }
+    ]);
+
+
+  function getDimensions(window, $textarea, computedStyle) {
+    var dimensions = {
+          width: window.parseInt(computedStyle.getPropertyValue('width'), 10),
+          height: window.parseInt(computedStyle.getPropertyValue('height'), 10),
+          minHeight: window.parseInt(computedStyle.getPropertyValue('min-height'), 10),
+          maxHeight: window.parseInt(computedStyle.getPropertyValue('max-height'), 10),
+          outerHeight: 0
+        },
+        lineHeight = window.parseInt(computedStyle.getPropertyValue('line-height'), 10);
+
+    if (computedStyle.getPropertyValue('box-sizing') === 'border-box') {
+      dimensions.outerHeight =
+        window.parseInt(computedStyle.getPropertyValue('padding-top'), 10) +
+        window.parseInt(computedStyle.getPropertyValue('padding-bottom'), 10) +
+        window.parseInt(computedStyle.getPropertyValue('border-top-width'), 10) +
+        window.parseInt(computedStyle.getPropertyValue('border-bottom-width'), 10);
+    }
```

Comment:
The goal here is to calculate the vertical padding + border. We could subtract the inner height from the full height, but there is no one-step way to get the former since both offsetHeight and clientHeight include padding in their return value. So, we have to use the computed style for padding. `border-top + border-bottom` can be replaced by `offsetHeight - clientHeight`, but it doesn't make a big difference (shorter, but also less straightforward).